### PR TITLE
GDB netproxy restructured as a shim

### PR DESCRIPTION
**Pull request by danallan**
This PR makes a major change to the GDB plugin's architecture by removing its netproxy in favor of a shim. It is changed in the following ways:
- rather than have a runner start gdbserver while the netproxy starts GDB (the latter of which must wait for compilation before connecting to gdbserver), the shim is responsible for managing both processes. This eliminates a number of complex synchronization and process management issues.
- this eliminates the need for a netproxy altogether, and the debugger plugin therefore communicates with the shim directly via a socket as supported by #38.
- the shim is bundled with the plugin (as the netproxy was), but on load the gdbdebugger plugin writes the shim to the workspace disk so that it is always available for a runner. The gdbdebugger is only registered as a debugger if the shim is written to disk successfully.
- the runners that rely on GDB as a debugger therefore start the shim directly, and not gdb or gdbserver (see https://github.com/c9/c9.ide.run/pull/16).

Given the change, the shim now acts as a standalone nodejs script, and accepts arguments to alter its configuration. The parsing/handling of these arguments in combination with shifting management of both gdb client and server process in the shim represent a good chunk of the line changes on this PR.

I also snuck in more generalized signal handling (in addition to just SIGSEGV) and a couple of minor other fixes to catch some common problems.

Ultimately, this should address a number of issues with regard to the GDB plugin's reliability, including both #13 and #15. 

**Important note:** 
1. This PR depends entirely on acceptance of #38 first.
2. Once accepted, this PR should be deployed at the same time as https://github.com/c9/c9.ide.run/pull/16 to ensure consistent access to the GDB debugger.

### CODE REVIEW

File: debuggers/gdb/gdbdebugger.js [hunk 1]
```diff
@@ -60,8 +58,20 @@ define(function(require, exports, module) {
                     ["autoshow", "true"]
                 ]);
             });
-
+            
+            // must register ASAP, or debugger won't be ready for reconnects
             debug.registerDebugger(TYPE, plugin);
+
+            // writeFile root is workspace directory, unless given ~
+            var shimPath = "~/bin/c9gdbshim.js";
```
Comment (nightwing) [inline debuggers/gdb/gdbdebugger.js:66]: I am looking for a way to make writefile call on demand, and only when missing
would it be ok to write this to `~/.c9/bin/c9gdbshim-<md5>.js` instead, or is this supposed to be used from other places too?

Comment (danallan) [inline debuggers/gdb/gdbdebugger.js:66]: MD5 or other versioning would be fine, so long as we have a symlink or some other mechanism that always points to the latest revision from a known path, e.g., `~/.c9/bin/c9gdbshim.js` is a link for `~/.c9/bin/c9gdbhim-<md5>.js`?

File: debuggers/gdb/shim.js [hunk 2]
```diff
@@ -232,9 +232,9 @@ function Executable() {
     this.spawn = function(callback) {
         var args = ["--once", ":"+GDB_PORT, BIN].concat(ARGS);
         this.proc = spawn("gdbserver", args, {
-            detached: false,
+            detached: true,
```
Comment (nightwing) [inline debuggers/gdb/shim.js:245]: why is detached flag needed? It seems like it has effect only when stdio is 'ignore'

**Comment by nightwing** [comment]
This is absolutely awesome! 

There are several small issues i need to resolve before releasing this, but they are mostly minor tweaks
- [x] remove hardcoded /home/ubuntu paths to allow this to work on ssh workspaces
- [x] show error when gdb or gdbserver are missing.
- [ ] test on windows (optional)
- [ ] find a way to not upload shim every time ide is loaded (nice).
- [ ] find a way to update old runners, that are modified by users, or show a clear warning that debugging won't work.

File: debuggers/gdb/shim.js (hunk 2 above)
Comment (danallan) [inline debuggers/gdb/shim.js:245]: Signals to the shim get distributed to all its child processes, unless they're detached into their own group. If the gdbserver was not detached, we lose control over signals. For example, create a C program with an infinite loop, like this:

``` c
int main(void) {
    int i = 0;
    while (1) { i = 1; }
}
```

Then debug, and when it's running hit ctrl-c. With `detached: true`, we can pause execution (as is expected with normal debugging in GDB; a second ctrl-c will stop debugging). With `detached: false`, SIGINT gets sent to gdbserver and the entire debug process is destroyed with only one ctrl-c.

**Comment by danallan** [comment]
Thanks, @nightwing! Nice list of finishing touches, too.
By the way, I had to make an additional PR just now (https://github.com/c9/c9.ide.run.debug/pull/40/commits/89b26d32ed9712384dfade75123c5e6f18e9fb69) as I realized I was needlessly unlinking the socketfile on shim exit (and occasionally interfered with the shim if a second process happened to use the socket while the first was exiting). Just thought I'd point it out in case you were making tweaks on a copy of the code.

**Pull request merged**